go/storage/mkvs/db/badger/badger.go

```diff
@@ -672,20 +773,69 @@ func (d *badgerNodeDB) Prune(ctx context.Context, version uint64) error {
 	return nil
 }
 
-func (d *badgerNodeDB) NewBatch(oldRoot node.Root, version uint64, chunk bool) api.Batch {
+func (d *badgerNodeDB) StartMultipartInsert(version uint64) error {
+	d.metaUpdateLock.Lock()
+	defer d.metaUpdateLock.Unlock()
+
+	if d.multipart {
+		if d.multipartVersion != version {
+			return api.ErrMultipartInProgress
+		}
+		// Multipart already initialized, so this was probably
+		// called e.g. as part of a further checkpoint restore.
+		return nil
+	}
+
+	tx := d.db.NewTransactionAt(tsMetadata, true)
+	defer tx.Discard()
+	if err := d.meta.setMultipartVersion(tx, version); err != nil {
+		return err
+	}
+	if err := tx.CommitAt(tsMetadata, nil); err != nil {
+		return err
+	}
+
+	d.multipart = true
+	d.multipartVersion = version
+
+	return nil
+}
+
+func (d *badgerNodeDB) AbortMultipartInsert() error {
+	d.metaUpdateLock.Lock()
+	defer d.metaUpdateLock.Unlock()
+
+	return d.cleanMultipartLocked(true)
+}
+
+func (d *badgerNodeDB) NewBatch(oldRoot node.Root, version uint64, chunk bool) (api.Batch, error) {
 	// WARNING: There is a maximum batch size and maximum batch entry count.
 	// Both of these things are derived from the MaxTableSize option.
 	//
 	// The size limit also applies to normal transactions, so the "right"
 	// thing to do would be to either crank up MaxTableSize or maybe split
 	// the transaction out.
 
-	return &badgerBatch{
-		db:      d,
-		bat:     d.db.NewWriteBatchAt(versionToTs(version)),
-		oldRoot: oldRoot,
-		chunk:   chunk,
+	d.metaUpdateLock.Lock()
+	defer d.metaUpdateLock.Unlock()
+	if d.multipart && d.multipartVersion != version {
+		return nil, api.ErrInvalidMultipartVersion
+	}
+
```

Comment:
Since this method can now return errors, you can move the read only check from the top of the `Commit` method here.